From 5d10dc5da2508530a38c66d4aa1ec234327a3b34 Mon Sep 17 00:00:00 2001
From: Ellen Poe <ellen.h.poe@gmail.com>
Date: Sun, 12 Feb 2023 07:34:55 -0800
Subject: [PATCH] Fix AudioStreamGenerator stopping playback after a buffer
 underrun

---
 servers/audio_server.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/servers/audio_server.cpp b/servers/audio_server.cpp
index 0344bf322d68..5e71d7af0550 100644
--- a/servers/audio_server.cpp
+++ b/servers/audio_server.cpp
@@ -354,7 +354,7 @@ void AudioServer::_mix_step() {
 			playback->stream_playback->tag_used_streams();
 		}
 
-		if (mixed_frames != buffer_size) {
+		if (!playback->stream_playback->is_playing()) {
 			// We know we have at least the size of our lookahead buffer for fade-out purposes.
 
 			float fadeout_base = 0.94;
