From 87fcee2d7fe34fe1ee1de462d56b7f4207672c02 Mon Sep 17 00:00:00 2001
From: David Snopek <dsnopek@gmail.com>
Date: Thu, 11 May 2023 16:58:52 -0500
Subject: [PATCH] Fix 'linux' and specific BSD feature tags

---
 platform/linuxbsd/os_linuxbsd.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/platform/linuxbsd/os_linuxbsd.cpp b/platform/linuxbsd/os_linuxbsd.cpp
index 82500f83cb58..c8e50e00769b 100644
--- a/platform/linuxbsd/os_linuxbsd.cpp
+++ b/platform/linuxbsd/os_linuxbsd.cpp
@@ -494,6 +494,11 @@ bool OS_LinuxBSD::_check_internal_feature_support(const String &p_feature) {
 		return true;
 	}
 
+	// Match against the specific OS (linux, freebsd, etc).
+	if (p_feature == get_name().to_lower()) {
+		return true;
+	}
+
 	return false;
 }
 
